src/backend/utils/mmgr/aset.c

```diff
@@ -998,6 +1011,118 @@ static void AllocSetReleaseAccountingForAllAllocatedChunks(MemoryContext context
 #endif
 }
 
+#ifdef EXTRA_DYNAMIC_MEMORY_DEBUG
+typedef struct
+{
+	AllocChunk chunk;
+	uint64_t count;
+	uint64_t bytes;
+} TopAllocChunkStats;
+
+static void update_AllocChunkStats(TopAllocChunkStats *saveStats, int *saveStatsCount,
+		AllocChunk chunk, uint64_t count, uint64_t bytes)
+{
+	int idx = *saveStatsCount;
+	for (; idx > 0; idx--)
+	{
+		if (bytes > saveStats[idx - 1].bytes)
+		{
+			if ((idx == *saveStatsCount && *saveStatsCount < MAX_TOP_ALLOC_CHUNK_STATS) ||
+			    idx < *saveStatsCount)
+			{
+				saveStats[idx] = saveStats[idx - 1];
+			}
+			continue;
+		}
+
+		// no space to save stats
+		if (*saveStatsCount == MAX_TOP_ALLOC_CHUNK_STATS && idx == MAX_TOP_ALLOC_CHUNK_STATS)
+			return;
+
+		break;
+	}
+
+	saveStats[idx].chunk = chunk;
+	saveStats[idx].count = count;
+	saveStats[idx].bytes = bytes;
+
+	if (*saveStatsCount < MAX_TOP_ALLOC_CHUNK_STATS)
+		(*saveStatsCount)++;
+}
```

Comment:
You might use heap-base sorting to define top-N items for log-linear time.

Reply:
For heap-base sorting we have to alloc big size of dynamic memory for full table of chunks (before it we have to calculate size of table). I wouldn't to alloc dynamic memory at function which prints dynamic alloctions table (I think that it is incorrect way, because it may screw up the waiting statistics).

Reply:
> For heap-base sorting we have to alloc big size of dynamic memory for full table of chunks

Why? Bounded heap sort requires just limit number of slots. You might read some rough [exploration](https://stackoverflow.com/questions/33936485/understanding-order-by-using-top-n-heapsort-in-postgres#answer-33939316) how the similar algorithm is implemented in postgres